### PR TITLE
Instrument tracing across dashboards

### PR DESCRIPTION
**Pull request by jalevin**
**What is this feature?**
This PR adds instruments pkg/services/dashboards and api/dashboards by adding spans to all methods that receive and propagate a context.

**Why do we need this feature?**
We're working on a lot of performance improvements for Grafana and having this information will help us identify bottlenecks

### CODE REVIEW

File: pkg/api/dashboard.go [hunk 1]
```diff
@@ -81,8 +85,9 @@ func dashboardGuardianResponse(err error) response.Response {
 // 404: notFoundError
 // 500: internalServerError
 func (hs *HTTPServer) GetDashboard(c *contextmodel.ReqContext) response.Response {
-	ctx, span := hs.tracer.Start(c.Req.Context(), "api.GetDashboard")
```
Comment (mildwonkey) [inline pkg/api/dashboard.go:84]: This one *is* GetDashboard, don't rename it!

**Review by mildwonkey — CHANGES_REQUESTED** [review]
I don't think we should be initializing tracers as package variables, that runs counter to our usual patterns and adds confusion. I'm also a little concerned about how some of these will look, such as `accesscontrol.acimpl.GetUserPermissions` -  I think the original name "authz.GetUserPermissionsOSS" was deliberate because that's _not_ the implementation of `accesscontrol.acimpl` that's used in cloud, we have an enterprise accesscontrol extension.

**Comment by jalevin** [comment]
> I don't think we should be initializing tracers as package variables, that runs counter to our usual patterns and adds confusion. 

The OTEL implementation for Go is a little confusing but it's calling GetTracer which is grabbing a global tracer. The existing pattern is quite verbose. We can avoid passing it through to every service if we do it this way. It will also clean up our tests.

> I'm also a little concerned about how some of these will look, such as `accesscontrol.acimpl.GetUserPermissions` - I think the original name "authz.GetUserPermissionsOSS" was deliberate because that's _not_ the implementation of `accesscontrol.acimpl` that's used in cloud, we have an enterprise accesscontrol extension.

I noticed that as well. The issue is that authz is ambiguous when trying to figure out which package it's in and GetUserPermissionsOSS isn't a function that exists. I used the form `folder.package.function` for specificity. I will rename in enterprise to delineate between the two

File: pkg/api/dashboard.go [hunk 2]
```diff
@@ -81,8 +85,9 @@ func dashboardGuardianResponse(err error) response.Response {
...
+	ctx, span := tracer.Start(c.Req.Context(), "api.isDashboardStarredByUser")
 	defer span.End()
+	c.Req = c.Req.WithContext(ctx)
```
Comment (dafydd-t) [inline pkg/api/dashboard.go:90]: Kristin's comment was resolved but I still see it - the span name here is inconsistent with the method name

**Review by dafydd-t — APPROVED** [review]
This looks good to me. The naming conventions are consistent with what you introduced in [this other PR](https://github.com/grafana/grafana/pull/91864).

File: pkg/api/dashboard.go (hunk 2 above)
Comment (jalevin) [reply pkg/api/dashboard.go]: I forgot to push! Thank you.

**Pull request merged**